### ISSUE TITLE
Training custom dataset on kangaroo

### ISSUE THREAD

**Issue by voqtuyen** [experiencor/basic-yolo-keras#203]
I followed the tutorial on how to train custom dataset and i chose kangaroo.
Here is the log:

```
31/50 [=================>............] - ETA: 3:33 - loss: 1.4822Dummy Line 	[0]
Loss XY 	[0.0825796127]
Loss WH 	[1.9374851]
Loss Conf 	[0.0127458191]
Loss Class 	[0]
Total Loss 	[2.03281069]
Current Recall 	[0]
Average Recall 	[0.164205089]
32/50 [==================>...........] - ETA: 3:21 - loss: 1.4994Dummy Line 	[0]
Loss XY 	[0.106220111]
Loss WH 	[1.57395577]
Loss Conf 	[0.00985390227]
Loss Class 	[0]
Total Loss 	[1.69002986]
Current Recall 	[0]
Average Recall 	[0.163336292]
33/50 [==================>...........] - ETA: 3:11 - loss: 1.5051Dummy Line 	[0]
Loss XY 	[0.0952992663]
Loss WH 	[1.51814735]
Loss Conf 	[0.00849177]
Loss Class 	[0]
Total Loss 	[1.62193835]
Current Recall 	[0]
Average Recall 	[0.162476614]
34/50 [===================>..........] - ETA: 2:59 - loss: 1.5086Dummy Line 	[0]
Loss XY 	[0.093166858]
Loss WH 	[1.39261317]
Loss Conf 	[0.0124569423]
Loss Class 	[0]
Total Loss 	[1.49823689]
Current Recall 	[0]
Average Recall 	[0.161625952]
```
The loss class is always 0, and the final model cannot predict the kangaroo correctly. Any suggestions on how to resolve it? Thanks so much!

**Comment by experiencor** [comment]
@voqtuyen Class loss is always 0 as you only have 1 class. You cannot get it wrong.

For correct prediction, please use Full Yolo and train more to reduce the loss further.

**Comment by voqtuyen** [comment]
Thanks @experiencor for your reply, i will try a re-training as your suggestion

**Comment by voqtuyen** [comment]
Hi @experiencor ,

I see that during training the recall decreases overtime? I think it's quite strange and it results in my wrong prediction result, is it supposed to be normal? Thanks for your time!

**Comment by experiencor** [comment]
Hi @voqtuyen, I find that the warmup stage is quite confusing so I changed the code to auto handle this case. The training process is now simply ```python train.py -c config.json```. May you try it again.

**Comment by voqtuyen** [comment]
Hi @experiencor,

Thanks for your quick reply, i will give it a try then!
Have a nice day!

**Comment by voqtuyen** [comment]
Hi @experiencor,
I have another problem here, i have trained kangaroo dataset with the final average recall about 0.4. The model is able to predict the kangaroo on images but with default 0.3 confidence, it think a person is a kangaroo in some cases. However, my real problem is that when i test the model on cat.jpg or dog.jpg... it also thinks a cat is a kangaroo. For now, i don't know why causes this to mis-classify these cases. 
Thanks for your help!

**Comment by experiencor** [comment]
@voqtuyen Yes, your result is normal to me. One way to address this issue to include non-kangaroo images in the dataset. You can get some from COCO too. The network will learn to recognize non-kangaroo and not just label every animal as a kangaroo.

**Comment by voqtuyen** [comment]
Hi @experiencor ,
Thanks for your useful info. The way to do this is to just put more non-kangaroo dataset like COCO to the images & annotations dir and re-run the training process. The preprocessing step will handle objects that don't belong to the predefined labels and leave all the x_batch, y_batch to zeros to denote the absence of kangaroo. Is that correct?

**Comment by experiencor** [comment]
@voqtuyen Yes. I combine the images and the annotations (put them in common folders). Then set labels in config.json to like ["kangaroo", "dog", "cat"]. It will learn a detector which can distinguish those three animals.

If you just want to detect kangaroo, you can make blank annotations for those additional images then put ["kangaroo"] in the labels settings.

**Comment by voqtuyen** [comment]
Thanks @experiencor,

I see that you divide the network into 2 parts: backend and frontend so that you can easily change the CNN architecture and also easily change the config of the last prediction layer, so my question is given a full model, how can i extract only the backend or frontend as you did? Thanks very much!

**Comment by experiencor** [comment]
@voqtuyen It's straight-forward but involves quite an effort. Make the backend submodel, make the frontend submodel, combine them to make the full model, load the weight to the full model, save each submodel separately. I did all these steps in the constructor of YOLO in frontend.py except the last two steps. You probably need a jupyter notebook to make your life easier.
